verification/utils/bitwise/bitwise-eqs.gobra

```diff
@@ -40,4 +40,101 @@ pure
 ensures 0 <= b & 0x7 && b & 0x7 <= 7
 ensures res == b & 0x7
 decreases
-func BitAnd7(b int) (res int)
\ No newline at end of file
+func BitAnd7(b int) (res int)
+
+ghost
+pure
+ensures 0 <= uint8(b >> 30) && uint8(b >> 30) < 4
+ensures res == uint8(b >> 30)
+decreases
+func First2Bits(b uint32) (res uint8)
+
+ghost
+pure
+requires b < 2147483648
+ensures  a >> 30 == (a | b) >> 30
+ensures  res == uint8(a >> 30)
+decreases
+func PreservesFirst2Bits(a uint32, b uint32) (res uint8)
+
+ghost
+pure
+requires 0 <= b && b <= 24
+ensures uint32(a & 0x3F) << b < 2147483648
+ensures res == uint32(a & 0x3F) << b
+decreases
+func LessThan2to31(a uint8, b int) (res uint32)
+
+ghost
+pure
+requires 0 <= b && b <= 16
+ensures 0 <= uint32(a) << b && uint32(a) << b < 2147483648
+ensures res == uint32(a) << b
+decreases
+func LessThan2to31_byte(a uint8, b int) (res uint32)
+
+ghost
+pure
+requires 0 <= a && a < 2147483648
+requires 0 <= b && b < 2147483648
+ensures 0 <= a | b && a | b < 2147483648
+ensures res == a | b
+decreases
+func LessThan2to31_bitwiseOr(a uint32, b uint32) (res uint32)
+
+ghost
+pure
+ensures a == a << 0
+ensures res == a
+decreases
+func ShiftBy0(a uint32) (res uint32)
+
+ghost
+pure
+ensures uint8(a >> 24) >> 6 == uint8(a >> 30)
+ensures res == uint8(a >> 30)
+decreases
+func ShiftBy24and6(a uint32) (res uint8)
+
+ghost
+pure
+ensures a | b == b | a
+ensures res == a | b
+decreases
+func BitOrCommutative(a uint32, b uint32) (res uint32)
+
+ghost
+pure
+ensures uint8(uint32(a) << 24) >> 30 == a >> 6
+ensures res == a >> 6
+decreases
+func ShiftBy24and30(a uint8) (res uint8)
+
+ghost
+pure
+requires 0 <= CurrINF && CurrINF < 4
+requires 0 <= SegLen0 && SegLen0 <= 0x3F
+requires 0 <= SegLen1 && SegLen1 <= 0x3F
+requires 0 <= SegLen2 && SegLen2 <= 0x3F
+ensures uint8(line >> 30) == CurrINF
+ensures uint8(line >> 24) >> 6 == CurrINF
+ensures uint8(line >> 12) & 0x3F == SegLen0
+ensures (uint8(line >> 8) >> 4 | (uint8(line >> 16) << 4)) & 0x3F == SegLen0
+ensures uint8(line >> 6) & 0x3F == SegLen1
+ensures (uint8(line >> 6) | uint8(line >> 8) << 2) & 0x3F == SegLen1
+ensures uint8(line) & 0x3F == SegLen2
+ensures line == uint32(CurrINF)<<30 | uint32(CurrHF & 0x3F)<<24 | uint32(SegLen0 & 0x3F)<<12 | uint32(SegLen1 & 0x3F)<<6 | uint32(SegLen2 & 0x3F)
+decreases
+func SerializingHdr(CurrINF, CurrHF, SegLen0, SegLen1, SegLen2 uint8) (line uint32)
+
+ghost
+pure
+requires line == uint32(raw3) | uint32(raw2)<<8 | uint32(raw1)<<16 | uint32(raw0)<<24
+ensures  uint8(line >> 30) == raw0 >> 6
+ensures  uint8(line >> 12) & 0x3F == (raw2 >> 4 | raw1 << 4) & 0x3F 
+ensures  uint8(line >> 6) & 0x3F == (raw3 >> 6 | raw2 << 2) & 0x3F
+ensures  uint8(line) & 0x3F == raw3 & 0x3F
+ensures  0 <= uint8(line >> 30) && uint8(line >> 30) < 4
+ensures  res == true
+decreases
+func DeserializingHdr(raw0, raw1, raw2, raw3 uint8, line uint32) (res bool)
```

Comment:
These theorems are way to hard to judge by themselves. I would split them. Otherwise, if it is useful for you to have a single large theorem like this, then I would split the subtheorems/lemmas into separate methods and I would make this function non-abstract, implemented via calls to simpler results.

Reply:
Also, for a bit of context, what I envision for these "axioms" is for them to be actually verified. To that end, I plan to introduce an annotation like the following
```go
ghost
ensures a == a << 0
ensures res == a
decreases
interpret-bitvec // <------ HERE
pure func ShiftBy0(a uint32) (res uint32) {
    // empty body should be work
}
``` 
the idea for `interpret-bitvec` is that any method annotated with it will actually encode bitvectors into smt bitvectors, and I am not sure how well performance will scale in these larger methods. If, instead, they are just calls to other intermediate results, then they don't even need the flag `interpret-bitvec`